contracts/rewards/src/contract.rs

```diff
@@ -26,21 +29,90 @@ pub fn instantiate(
         },
     )?;
 
+    PARAMS.save(
+        deps.storage,
+        &StoredParams {
+            params: msg.params,
+            last_updated: Epoch {
+                epoch_num: 0,
+                block_height_started: env.block.height,
+            },
+        },
+    )?;
+
     Ok(Response::new())
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
 pub fn execute(
     deps: DepsMut,
-    _env: Env,
-    _info: MessageInfo,
+    env: Env,
+    info: MessageInfo,
     msg: ExecuteMsg,
 ) -> Result<Response, axelar_wasm_std::ContractError> {
-    let mut _contract = Contract::new(deps);
     match msg {
-        ExecuteMsg::RecordParticipation { .. } => todo!(),
-        ExecuteMsg::AddRewards { .. } => todo!(),
-        ExecuteMsg::DistributeRewards { .. } => todo!(),
-        ExecuteMsg::UpdateParams { .. } => todo!(),
+        ExecuteMsg::RecordParticipation {
+            event_id,
+            worker_address,
+        } => {
+            let worker_address = deps.api.addr_validate(&worker_address)?;
+            Contract::new(deps)
+                .record_participation(
+                    nonempty::String::try_from(event_id)
+                        .change_context(ContractError::InvalidEventId)?,
+                    worker_address,
+                    info.sender,
+                    env.block.height,
+                )
+                .map_err(axelar_wasm_std::ContractError::from)?;
+
+            Ok(Response::new())
+        }
+        ExecuteMsg::AddRewards { contract_address } => {
+            let contract_address = deps.api.addr_validate(&contract_address)?;
+            let mut contract = Contract::new(deps);
+            let amount = info
+                .funds
+                .iter()
+                .find(|coin| coin.denom == contract.config.rewards_denom)
+                .filter(|_| info.funds.len() == 1)
+                .ok_or(ContractError::WrongDenom)?
+                .amount;
+
+            contract.add_rewards(
+                contract_address,
+                nonempty::Uint128::try_from(amount).change_context(ContractError::ZeroRewards)?,
+            )?;
+
+            Ok(Response::new())
+        }
+        ExecuteMsg::DistributeRewards {
+            contract_address,
+            epoch_count,
+        } => {
+            let contract_address = deps.api.addr_validate(&contract_address)?;
+            let mut contract = Contract::new(deps);
+            let rewards = contract
+                .distribute_rewards(contract_address, env.block.height, epoch_count)
+                .map_err(axelar_wasm_std::ContractError::from)?;
+
+            let mut msgs = vec![];
+            for (addr, amount) in rewards {
+                msgs.push(BankMsg::Send {
+                    to_address: addr.into(),
+                    amount: vec![Coin {
+                        denom: contract.config.rewards_denom.clone(),
+                        amount,
+                    }],
+                })
+            }
```

Comment:
```suggestion
            let msgs = rewards.into_iter().map(|(addr, amount)| BankMsg::Send {
                to_address: addr.into(),
                amount: vec![Coin {
                    denom: contract.config.rewards_denom.clone(),
                    amount,
                }],
            });
```
Avoid mutation whenever possible. Now, I can be wrong but I don't think iterating a hash map is necessarily deterministic in terms of ordering. You may have to switch to a BTreeMap instead.